serverless/lambda/src/main/java/uk/gov/di/lambdas/LogoutHandler.java

```diff
@@ -4,29 +4,164 @@
+    public LogoutHandler(
+            ConfigurationService configurationService,
+            SessionService sessionService,
+            DynamoClientService dynamoClientService) {
         this.configurationService = configurationService;
+        this.sessionService = sessionService;
+        this.dynamoClientService = dynamoClientService;
     }
 
     @Override
     public APIGatewayProxyResponseEvent handleRequest(
             APIGatewayProxyRequestEvent input, Context context) {
+        Optional<String> state = Optional.ofNullable(input.getQueryStringParameters().get("state"));
+        Optional<Session> sessionFromSessionCookie =
+                sessionService.getSessionFromSessionCookie(input.getHeaders());
+        return sessionFromSessionCookie
+                .map(t -> processLogoutRequest(t, input, state))
+                .orElse(generateDefaultLogoutResponse(state));
+    }
+
+    private APIGatewayProxyResponseEvent processLogoutRequest(
+            Session session, APIGatewayProxyRequestEvent input, Optional<String> state) {
+        Map<String, String> queryStringParameters = input.getQueryStringParameters();
+        Optional<CookieHelper.SessionCookieIds> sessionCookieIds =
+                CookieHelper.parseSessionCookie(input.getHeaders());
+
+        if (!session.getClientSessions().containsKey(sessionCookieIds.get().getClientSessionId())) {
+            throw new RuntimeException(
+                    format(
+                            "Client Session ID does not exist in Session: %s",
+                            session.getSessionId()));
+        }
+        if (!queryStringParameters.containsKey("id_token_hint")
+                || queryStringParameters.get("id_token_hint").isBlank()) {
+            sessionService.deleteSessionFromRedis(session.getSessionId());
+            return generateDefaultLogoutResponse(state);
+        }
+        if (!doesIDTokenExistInSession(queryStringParameters.get("id_token_hint"), session)) {
+            throw new RuntimeException(
+                    format("ID Token does not exist for Session: %s", session.getSessionId()));
+        }
+        if (!isIDTokenSignatureValid(
+                queryStringParameters.get("id_token_hint"), session.getSessionId())) {
+            throw new RuntimeException(
+                    format(
+                            "Unable to validate ID token signature for Session: %s",
+                            session.getSessionId()));
+        }
+        try {
+            String idTokenHint = queryStringParameters.get("id_token_hint");
+            SignedJWT idToken = SignedJWT.parse(idTokenHint);
+            Optional<String> audience =
+                    idToken.getJWTClaimsSet().getAudience().stream().findFirst();
+            sessionService.deleteSessionFromRedis(session.getSessionId());
+            return audience.map(
+                            t -> {
+                                final ClientRegistry clientRegistry;
```

Comment:
Can we pull this lambda out as a method instead? It feels a bit too large to inline.

Reply:
Done